### ISSUE TITLE
[Proposal] Reducing contact with the schema for Prisma Client users

### ISSUE THREAD

**Issue by madebysid** [prisma/prisma#2573]
Persisting our [Slack](https://prisma-company.slack.com/archives/CFM8D5VHC/p1590481767010100) conversation here, and giving my initial idea a bit more thought.

Also note that this issue assumes you're using Client & *only* Client. The conversation becomes irrelevant if you want to use Migrate.

## Problem:

For someone just starting out in the Prisma ecosystem, the schema is "one more thing to learn". I believe it would be beneficial to hide the schema entirely from people willing to use only Prisma Client, in the name of optional complexity. The schema would obviously still exist, but I'm trying to prevent people from having to deal with it until they absolutely have to / want to.

Currently, if you use *only* Prisma Client (for example with Nexus), your workflow might look something like this:
1. Run `prisma init` in your directory, this will create a `prisma/schema.prisma` file.
2. Edit this file's `datasource` block to point to the database you want to connect to
3. Run `prisma introspect` to populate the schema with models & fields from your database
4. Run `prisma generate` to generate Prisma Client
5. When and if your database's structure changes, repeat steps 3 - 5.

## Solution:

If someone's just using Client, they're likely not changing (or even looking at) the prisma schema. In this case, they do not really need to know it exists. All they're trying to do is get a type safe Client generated.
Here's the workflow I'm imagining:
1. Run `prisma init --db "postgres://sid@localhost:5432/database"` (Or more likely: `prisma init --db "DB_URL"`, aka using an environment variable)
2. When and if you database's structure changes, repeat step 1.

Here's what's happening under the hood:
1. Prisma generates a barebones schema, just like today's `prisma init` does in `node_modules/.prisma/schema.prisma`
2. Prisma introspects your database, and writes to `node_modules/.prisma/schema.prisma`
3. Prisma generates Prisma Client to `node_modules/@prisma/client`

Now, when people do want to mess with the schema (rather than have Prisma manage it), they can manually create a `./prisma/schema.prisma` file in their directory and everything else continues just like it does today. In other words, `prisma init` respects a schema file, if it exists. If it doesn't, it creates one under the hood. That's what this entire issue boils down to.

## Known Problems:

- How this would work on CI systems is unclear to me. In order for it to work, we'd need a way to persist their database url on version control somewhere. What better place to store it than the schema file? Aaand that kinda brings us back to square one.

- I can't help but notice that this is very similar to the init experience we had during the alpha days, and there were reasons we decided to discontinue it, some of them might apply here too.

Assuming we decide that such a thing even makes sense, it needs more thought, which is why this issue exists. Suggestions welcome!

**Comment by giraffesyo** [comment]
> How this would work on CI systems is unclear to me. In order for it to work, we'd need a way to persist their database url on version control somewhere. What better place to store it than the schema file? Aaand that kinda brings us back to square one.

What would be the problem with `primsa init` pulling the `DATABASE_URL` from the environment if no `--db` flag was passed? You wouldn't have trouble with CI this way, because you can set the environment variables on your CI server. CI can rerun the init every time if necessary.

**Comment by madebysid** [comment]
@giraffesyo Looking back at this, I don't think I explained what I was thinking very well. My "problem" with running `prisma init` every time on the CI is the fact that the database itself could change between runs (without you making any changes). This would mean the schema changes between runs, which means the generated client changes between runs.

That being said, I agree with you, pulling the DB URL from an env var & running init on every CI run is an acceptable thing to do, assuming that Typescript will likely complain if the generated client changed.

**Comment by thebiglabasky** [comment]
Thanks for the suggestion!

After looking into that further and discussing it with the team we decided not to look further down that path. Reasons:
 * The schema has been voiced as bringing value on its own, serving as a central place to understand the schema, in a more human-readable way than a SQL DDL. It serves as documentation for people onboarding on a project for example. This can be specifically interesting in cases with introspection
 * The workflows using migrate lean on the schema being a first-class citizen. Hiding the schema away would make no sense for these users. Moreover, making strong distinctions in the experience depending on which flow someone starts brings complexity (to the product, to the docs) and potentially more confusion. This could otherwise favor certain users versus others.

In a nutshell, while it's a Prisma-only concept to learn indeed, it's made in such a way that it's worth learning it. This learning curve is made easy through examples and the VS Code plugin. We'll stick to that.
I'll thus close the issue.
